sdks/python/apache_beam/runners/common.py

```diff
@@ -842,29 +847,37 @@ def _invoke_process_per_window(self,
 
   def try_split(self, fraction):
     # type: (...) -> Optional[Tuple[SplitResultPrimary, SplitResultResidual]]
-    if self.threadsafe_restriction_tracker and self.current_windowed_value:
+    if not self.is_splittable:
+      return None
+
+    with self.splitting_lock:
+      # Make a local reference to member variables that change references during
+      # processing under lock before attempting to split so we have a consistent
+      # view of all the references.
+      current_windowed_value = self.current_windowed_value
+      threadsafe_restriction_tracker = self.threadsafe_restriction_tracker
+      threadsafe_watermark_estimator = self.threadsafe_watermark_estimator
+
+    if threadsafe_restriction_tracker:
```

Comment:
Would it be better to both check `threadsafe_restriction_tracker ` and `threadsafe_watermark_estimator` for easy reading?

Reply:
There is an assumption that if one is set then the other is set. Similarly, if one is unset then the other is unset.

Reply:
I would add a comment above to state the assumption. But it's minor though.

Reply:
I have a much larger rewrite here to support per window invocation where this will become a non-issue so I'll pass on this since we will need to hold the lock for the entire split call.